<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Définition-des-machines-à-analyser-et-de-la-notion-d'arrêt" data-toc-modified-id="Définition-des-machines-à-analyser-et-de-la-notion-d'arrêt-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Définition des machines à analyser et de la notion d'arrêt</a></span><ul class="toc-item"><li><span><a href="#Machines" data-toc-modified-id="Machines-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Machines</a></span></li><li><span><a href="#Processus-et-exécution-des-instructions" data-toc-modified-id="Processus-et-exécution-des-instructions-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Processus et exécution des instructions</a></span></li><li><span><a href="#Fin-des-programmes" data-toc-modified-id="Fin-des-programmes-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Fin des programmes</a></span></li></ul></li><li><span><a href="#Générations-des-machines-et-analyse-globale-de-l'arrêt" data-toc-modified-id="Générations-des-machines-et-analyse-globale-de-l'arrêt-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Générations des machines et analyse globale de l'arrêt</a></span><ul class="toc-item"><li><span><a href="#Génération" data-toc-modified-id="Génération-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Génération</a></span></li><li><span><a href="#Analyse-d'un-programme" data-toc-modified-id="Analyse-d'un-programme-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Analyse d'un programme</a></span></li><li><span><a href="#Analyse-de-plusieurs-programmes-(tous)" data-toc-modified-id="Analyse-de-plusieurs-programmes-(tous)-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Analyse de plusieurs programmes (tous)</a></span></li></ul></li><li><span><a href="#Autres-analyses" data-toc-modified-id="Autres-analyses-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Autres analyses</a></span><ul class="toc-item"><li><span><a href="#Analyses-aléatoires" data-toc-modified-id="Analyses-aléatoires-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Analyses aléatoires</a></span></li><li><span><a href="#Analyses-plus-poussée" data-toc-modified-id="Analyses-plus-poussée-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Analyses plus poussée</a></span></li></ul></li><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Conclusion</a></span><ul class="toc-item"><li><span><a href="#Récapitulatif" data-toc-modified-id="Récapitulatif-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Récapitulatif</a></span></li><li><span><a href="#Variantes" data-toc-modified-id="Variantes-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Variantes</a></span></li><li><span><a href="#Rappels" data-toc-modified-id="Rappels-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Rappels</a></span></li></ul></li></ul></div>

# Problème de l'arrêt pour des machines jouets minimalistes finies

Le problème de l'arrêt consiste à déterminer si l'on peut, en toute généralité, écrire un programme A (comme "programme qui détermine l'Arrêt") qui pourra prendre en entrée un autre programme X et une donnée D (pour le programme X) et décide si le programme X va s'arrêter sur la donnée D. Le résultat classique du problème de l'arrêt conclue qu'il n'est pas possible de de produire un programme qui puisse dire en toute généralité si un autre programme va s'arréter.

La preuve est jolie, elle repose sur une astuce et un raisonnement par l'absurde (voir [wkpd](https://fr.wikipedia.org/wiki/Probl%C3%A8me_de_l%27arr%C3%AAt)). Mais elle n'est pas valide pour tous les cas. Pour les machines de Turing, la preuve est valide (heureusement). Mais les machines de Turing ne sont pas des machines réelles, en particulier, elles supposent l'existence d'une mémoire infinie. Non pas très grande mais infinie. Et la preuve du problème de l'arrêt utilise implicitement cet infini. Nos machines n'ont malheureusement pas une mémoire infinie ...

*Remarque : l'existence de programmes résolvant le problème de l'arrêt pour des machines finies est clair, mais théorique. Dans la pratique, dans le cas général (fini), les approches usuelles sont limitées par la combinatoire et ne peuvent résoudre le problème de l'arrêt que pour de petites machines. Dans la pratique le problème de l'arrêt subsiste.*

Pour des machines M ayant une mémoire finie (taille max fixée : T), le problème de l'arrêt ne tient pas, et il existe des programmes A (avec un peu plus de mémoire que les machines M \[donc A ne pourra pas s'appliquer à A {ce qui interdit la preuve classique du problème de l'arrêt}]) permettant de déterminer si les machines M s'arrêtent. Dans le cas présent, nous supposerons que l'on possède une super-machine avec une mémoire un peu supérieur à 2*T pour pouvoir faire fonctionner le programme A (à suivre).

## Définition des machines à analyser et de la notion d'arrêt

Pour la suite nous utiliserons des machines élémentaires ayant une architecture de Harvard avec une taille pour les programmes et les données fixe (aussi grande \[en théorie] que l'on veut, en pratique, aussi grande que nous permet la machine sur laquelle on va faire tourner notre programme d'analyse, ce qui risque d'être assez petit, désolé).

### Machines

Dans un premier temps, le langage machine autorisé pour ces machines comporte seulement les instructions suivantes :
* ADD Rg, Rd : Rg <= Rg + Rd (et mise à jour de Z en fonction du résultat de l'addition)
* BNE Adr : Si -Z, PC <= Adr

(attention, par défaut, l'instruction vide est traduite par "ADD R0, R0" qui n'est pas une instruction neutre)

Les tailles des instructions, des registres, des adresses doivent être compatibles. Par exemple, en prenant :

In [63]:
TailleMemoire = 7 #mots mémoire
TailleMemoireProgramme = 3 #mots mémoire
TailleMemoireDonnees = 4 #mots mémoire
TailleMotMemoire = 5 #bits
TailleBancRegistres = 4 #registres
TailleMotRegistre = 5 #bits
TailleInstruction = 5 #bits

* un codage binaire des instructions sur 1 bit : Add (0), BNE (1), et 
* 4 registres (numéro de registre codable sur 2 bits, taille du registre au moins la taille d'une adresse, d'une valeur ou d'une instruction => 5 bits), 
* 7 mots mémoire (adresse codable sur 3 bits), mots à répartir entre instructions du programme et données, par exemple, 3 mots de programme, 4 mots de données (architecture de Harvard)
x
Cela peut signifier une instruction sur 5 bits, soit pour la mémoire totale :
* 35 bits de mémoire au maximum (15 bits de programme \[environ une dizaine de milliers de programmes possibles] et 20 bits de données \[environ un million de données possibles])
* 16 (add) + 7 (bne) = 23 instructions différentes et donc 23**TailleMemoireProgramme  = 12 167 programmes différents

(*une variante plus petite peut être définie, avec moins de mémoire ou moins de registre, etc. le choix de la version proposée vient de ce qu'elle permet de réaliser un algorithme naïf de multiplication \[au moins].*)

In [64]:
NbInstructionsPossibles = TailleBancRegistres**2+TailleMemoire
NbProgrammesPossibles = NbInstructionsPossibles**TailleMemoireProgramme
NbDonneesPossibles = (1<<TailleMotMemoire)**TailleMemoireDonnees
NbMachinesPossibles = NbProgrammesPossibles*NbDonneesPossibles
print("Nb d'instructions différentes possibles : ",NbInstructionsPossibles,
      "\nNb de programmes différents possibles : ",NbProgrammesPossibles,
      "\nNb de données différentes possibles : ",NbDonneesPossibles,
      "\nNb de machines différentes possibles : ",NbMachinesPossibles)

Nb d'instructions différentes possibles :  23 
Nb de programmes différents possibles :  12167 
Nb de données différentes possibles :  1048576 
Nb de machines différentes possibles :  12758024192


### Processus et exécution des instructions

le processus (programme à l'exécution) sera déterminé par :
* la mémoire (programme et données)
* les valeurs des registres (PC, R0, R1, R2, R3)
* la valeur de Z

A l'initialisation, la mémoire étant fournie (programme et données) :
* Z = 0
* PC = 0
* R0, R1, R2, R3 prennent les valeurs en début de zone mémoire (partie donnée), R0=Mem[3], R1=Mem[4], R2=Mem[5], R3=Mem[6] 

In [65]:
def initProcessus(P,M):
    P["Memoire"] = M #TODO:faire une recopie mot à mot, avec contrôle <> "" et modulo TailleMotMemoire
    P["Z"] = 0
    P["PC"] = 0
    P["R0"] = M[TailleMemoireProgramme]&((1<<TailleMotRegistre)-1)
    P["R1"] = M[TailleMemoireProgramme+1]&((1<<TailleMotRegistre)-1)
    P["R2"] = M[TailleMemoireProgramme+2]&((1<<TailleMotRegistre)-1)
    P["R3"] = M[TailleMemoireProgramme+3]&((1<<TailleMotRegistre)-1)

Exemple, pour une mémoire de 8 mots de 5 bits : la multiplication avec 2 et 3 en entrées, résultat dans R3 :

In [66]:
MemoireTest = [["ADD","R3","R1"],["ADD","R0","R2"],["BNE",0],2,3,-1,0]
ProcessusTest = {}
initProcessus(ProcessusTest,MemoireTest)
ProcessusTest

{'Memoire': [['ADD', 'R3', 'R1'],
  ['ADD', 'R0', 'R2'],
  ['BNE', 0],
  2,
  3,
  -1,
  0],
 'Z': 0,
 'PC': 0,
 'R0': 2,
 'R1': 3,
 'R2': 31,
 'R3': 0}

### Fin des programmes

L'exécution d'un processus s'effectue, instruction après instruction, jusqu'à erreur ou fin de programme (arrivée à la fin de la zone de programme) ou boucle infinie.

In [67]:
def ExecuteUneInstruction(P):
    Mnemo = P["Memoire"][P["PC"]][0]
    if Mnemo=="ADD" :
        Res = P[P["Memoire"][P["PC"]][1]]+P[P["Memoire"][P["PC"]][2]]
        Res = Res & ((1<<TailleMotRegistre)-1)
        if Res == 0:
            P["Z"] = 1
        else :
            P["Z"] = 0            
        P[P["Memoire"][P["PC"]][1]]= Res
        P["PC"]=P["PC"]+1
        if P["PC"]==TailleMemoireProgramme :
            return "Fin de programme"
    elif Mnemo=="BNE" :
        if P["Z"]==0 :
            P["PC"]=P["Memoire"][P["PC"]][1]
            if ((P["PC"]<0) or (P["PC"]>=TailleMemoireProgramme)) :
                return "Erreur : Suite de Programme invalide"
        else :
            P["PC"]=P["PC"]+1
            if P["PC"]==TailleMemoireProgramme :
                return "Fin de programme"
    else :
        return "Erreur : Instruction invalide"

La fin de programme normale et la fin de programme en erreur sont détectées lors de l'exécution d'une instruction. Pour détecter l'exécution d'une boucle infinie, l'état du processus est inspecté selon la méthode de Floyd pour trouver les cycles dans les suites Un=f(Un-1) (voir wkpd) en exécutant le processus à deux vitesses différentes (le processus tortue : une instruction à la fois ; le processus lièvre : deux instructions à la suite à chaque fois), si la tortue et le lièvre se rencontrent c'est qu'il y a une boucle infinie.

In [ ]:
def MemeEtat(P,Q):
    if P["PC"]!=Q["PC"]:
        return False
    elif P["R0"]!=Q["R0"]:
        return False
    elif P["R1"]!=Q["R1"]:
        return False
    elif P["R2"]!=Q["R2"]:
        return False
    elif P["R3"]!=Q["R3"]:
        return False
    elif P["Z"]!=Q["Z"]:
        return False
    else :
        return True

In [ ]:
def ExecuteProcessus(P):
    Tortue = P.copy()
    Lievre = P.copy()
    continuation = True
    while continuation :
        ExecuteUneInstruction(Tortue)
        msg = ExecuteUneInstruction(Lievre)
        if msg is None :
            msg = ExecuteUneInstruction(Lievre)
        if msg :
            return msg
        elif MemeEtat(Tortue,Lievre) :
            return "Erreur : Boucle infinie"

In [68]:
def AnalyseMachine(M):
    P = {}
    initProcessus(P,M)
    return ExecuteProcessus(P)

L'analyse de la machine s'effectue en fonction de la mémoire :

In [69]:
# Choisir une machine pour l'analyse :
AnalyseMachine([["ADD","R3","R1"],["ADD","R0","R2"],["BNE",0],2,3,-1,0]) #test multiplication
#AnalyseMachine[["ADD","R3","R1"],["ADD","R0","R2"],["BNE",0],2,3,0,0]) # boucle infinie
#AnalyseMachine[["ADD","R3","R1"],["ADD","R0","R2"],["BNE",TailleMemoireProgramme+2],2,3,-1,0]) # erreur


'Fin de programme'

## Générations des machines et analyse globale de l'arrêt

### Génération

Une machine sera déterminée par son programme (parmi les 12167 programmes possibles) et ses données (parmi les 1048576 données possibles) :

In [ ]:
def GenereInstruction(NInstruction) :
    if NInstruction<(TailleBancRegistres**2) :
        return["ADD","R"+str((NInstruction&3)),"R"+str((NInstruction>>2)&3)] #2 bits par 2 bits, masque 11 = 3
    else :
        return ["BNE",NInstruction-(TailleBancRegistres**2)]

In [70]:
def GenereMachine(NProgramme,NDonnees) :
    return [GenereInstruction((NProgramme // (NbInstructionsPossibles ** 0))%NbInstructionsPossibles),
            GenereInstruction((NProgramme // (NbInstructionsPossibles ** 1))%NbInstructionsPossibles),
            GenereInstruction((NProgramme // (NbInstructionsPossibles ** 2))%NbInstructionsPossibles),
            (NDonnees >> (0*TailleMotMemoire)) & ((1<<TailleMotMemoire)-1),
            (NDonnees >> (1*TailleMotMemoire)) & ((1<<TailleMotMemoire)-1),
            (NDonnees >> (2*TailleMotMemoire)) & ((1<<TailleMotMemoire)-1),
            (NDonnees >> (3*TailleMotMemoire)) & ((1<<TailleMotMemoire)-1)]

Pour analyser un programme et une donnée particulière, il suffit de construire le processus associé et de l'exécuter :

In [71]:
from random import randint
#choisir une mémoire aléatoire ou particuliere :
MemoireExemple=GenereMachine(randint(0,NbProgrammesPossibles),randint(0,NbDonneesPossibles))
#MemoireExemple = [["ADD","R3","R1"],["ADD","R0","R2"],["BNE",0],2,3,-1,0]
AnalyseMachine(MemoireExemple)

'Erreur : Suite de Programme invalide'

### Analyse d'un programme

Pour analyser un programme pour toutes les données possible, il suffit de lancer l'analyse sur toutes les données :

In [72]:
def AnalyseProgramme(NP) :
    NbOk = 0
    NbOO = 0
    NbEr = 0
    ProcessusAnalyse = {}
    for i in range((1<<TailleMotMemoire)**TailleMemoireDonnees):
        MemoireAnalyse=GenereMachine(NP,i)
        initProcessus(ProcessusAnalyse,MemoireAnalyse)
        res=ExecuteProcessus(ProcessusAnalyse)
        if res == "Fin de programme" :
            NbOk = NbOk+1
        elif res == "Erreur : Boucle infinie" :
            NbOO = NbOO +1
        elif res == "Erreur : Suite de Programme invalide" : 
            NbEr = NbEr +1
        else :
            return res
    print ("Nb Ok :",NbOk)
    print ("Nb OO :",NbOO)
    print ("Nb Er :",NbEr)  

Par exemple (attention, cela peut prendre un peu de temps) :

In [73]:
from random import randint
NbTmp = randint(0,NbProgrammesPossibles)
print (NbTmp)
MemoireTmp=GenereMachine(NbTmp,0)
print (MemoireTmp)
#lancer l'analyse:
#AnalyseProgramme(NbTmp)

4492
[['ADD', 'R3', 'R1'], ['ADD', 'R3', 'R2'], ['ADD', 'R0', 'R2'], 0, 0, 0, 0]


Pour analyser un programme particulier, on peut prendre exemple sur AnalyseProgramme, par exemple pour analyser la multiplication précédente, en faisant varier les entiers à multiplier :

In [74]:
def AnalyseProgrammeMultiplication() :
    NbOk = 0
    NbOO = 0
    NbEr = 0
    MemoireTri = [["ADD","R3","R1"],["ADD","R0","R2"],["BNE",0],"operande1","operande2",-1,0]
    ProcessusTri = {}
    for i in range(1<<TailleMotMemoire):
        for j in range(1<<TailleMotMemoire):
            MemoireTri[3]=i
            MemoireTri[4]=j            
            initProcessus(ProcessusTri,MemoireTri)
            res=ExecuteProcessus(ProcessusTri)
            if res == "Fin de programme" :
                NbOk = NbOk+1
            elif res == "Erreur : Boucle infinie" :
                NbOO = NbOO +1
            elif res == "Erreur : Suite de Programme invalide" : 
                NbEr = NbEr +1
            else :
                return res
    print ("Nb Ok :",NbOk)
    print ("Nb OO :",NbOO)
    print ("Nb Er :",NbEr)  

In [75]:
#lancer l'analyse:
AnalyseProgrammeMultiplication()

Nb Ok : 1024
Nb OO : 0
Nb Er : 0


### Analyse de plusieurs programmes (tous)

Pour analyser tous les programmes possibles sur toutes les données, cela risque de prendre du temps, mais c'est facile à programmer (rappel, pour analyser un programme particulier sur toutes les données : AnalyseProgramme(NProg), pour analyser une machine particulier, i.e. un programme et ses données : AnalyseMachine(Memoire)) :

In [76]:
def AnalyseDeA(NPDebut,NPFin) :
    print ("Analyse des machines")
    print ("De :",NPDebut)
    print ("Jusqu'à :",NPFin)  
    print("-------------")    
    NbOkTot = 0
    NbOOTot = 0
    NbErTot = 0
    ProcessusAnalyse = {}
    for m in range(NPDebut,NPFin) :
        NbOk = 0
        NbOO = 0
        NbEr = 0
        for i in range((1<<TailleMotMemoire)**TailleMemoireDonnees):
            MemoireAnalyse=GenereMachine(m,i)
            initProcessus(ProcessusAnalyse,MemoireAnalyse)
            res=ExecuteProcessus(ProcessusAnalyse)
            if res == "Fin de programme" :
                NbOk = NbOk+1
            elif res == "Erreur : Boucle infinie" :
                NbOO = NbOO +1
            elif res == "Erreur : Suite de Programme invalide" : 
                NbEr = NbEr +1
            else :
                return res
        print (m)
        NbOkTot = NbOk+NbOkTot
        NbOOTot = NbOO+NbOOTot
        NbErTot = NbEr+NbErTot        
        print ("Nb Ok :",NbOk,"\tNb Ok total :",NbOkTot)
        print ("Nb OO :",NbOO,"\tNb OO total :",NbOOTot)
        print ("Nb Er :",NbEr,"\tNb Er total :",NbErTot)  
        print("-------------")

L'analyse de l'ensemble des programmes possibles, avec toutes les données possibles est donc maintenant facile (mais cela va prendre du temps) :

In [77]:
#Pour lancer l'analyse globale :
#AnalyseDeA(0,NbProgrammesPossibles)

## Autres analyses

### Analyses aléatoires

Pour lancer une analyse de quelques machines prises aléatoirement avec des données elle-même prises au hasard (cela sera utile quand les machines et la taille de la mémoire sera plus grande et que l'on ne pourra plus analyser toutes les machines sur toutes les données) :

In [78]:
def AnalyseAlea(NbMachines,NbDonnees) :
    print ("Analyse de ",NbMachines,"machines, chacune sur ",NbDonnees,"données")
    print("-------------")    
    NbOkTot = 0
    NbOOTot = 0
    NbErTot = 0
    ProcessusAnalyse = {}
    for m in range(NbMachines) :
        NbOk = 0
        NbOO = 0
        NbEr = 0
        NProgrammeAleatoire = randint(0,NbProgrammesPossibles)
        for i in range(NbDonnees):
            MemoireAnalyse=GenereMachine(NProgrammeAleatoire,randint(0,NbDonneesPossibles))
            initProcessus(ProcessusAnalyse,MemoireAnalyse)
            res=ExecuteProcessus(ProcessusAnalyse)
            if res == "Fin de programme" :
                NbOk = NbOk+1
            elif res == "Erreur : Boucle infinie" :
                NbOO = NbOO +1
            elif res == "Erreur : Suite de Programme invalide" : 
                NbEr = NbEr +1
            else :
                return res
        print ("Programme : ",NProgrammeAleatoire)
        print (GenereMachine(NProgrammeAleatoire,0))
        NbOkTot = NbOk+NbOkTot
        NbOOTot = NbOO+NbOOTot
        NbErTot = NbEr+NbErTot        
        print ("Nb Ok :",NbOk,"\tNb Ok total :",NbOkTot,"\tNb Ok total :",100*NbOkTot/(NbOkTot+NbOOTot+NbErTot),"%")
        print ("Nb OO :",NbOO,"\tNb OO total :",NbOOTot,"\tNb OO total :",100*NbOOTot/(NbOkTot+NbOOTot+NbErTot),"%")
        print ("Nb Er :",NbEr,"\tNb Er total :",NbErTot,"\tNb Er total :",100*NbErTot/(NbOkTot+NbOOTot+NbErTot),"%")  
        print("-------------")

In [79]:
AnalyseAlea(10,100)

Analyse de  10 machines, chacune sur  100 données
-------------
Programme :  3824
[['ADD', 'R2', 'R1'], ['ADD', 'R1', 'R1'], ['ADD', 'R3', 'R1'], 0, 0, 0, 0]
Nb Ok : 100 	Nb Ok total : 100 	Nb Ok total : 100.0 %
Nb OO : 0 	Nb OO total : 0 	Nb OO total : 0.0 %
Nb Er : 0 	Nb Er total : 0 	Nb Er total : 0.0 %
-------------
Programme :  6552
[['BNE', 4], ['ADD', 'R0', 'R2'], ['ADD', 'R0', 'R3'], 0, 0, 0, 0]
Nb Ok : 0 	Nb Ok total : 100 	Nb Ok total : 50.0 %
Nb OO : 0 	Nb OO total : 0 	Nb OO total : 0.0 %
Nb Er : 100 	Nb Er total : 100 	Nb Er total : 50.0 %
-------------
Programme :  11042
[['ADD', 'R2', 'R0'], ['BNE', 4], ['BNE', 4], 0, 0, 0, 0]
Nb Ok : 2 	Nb Ok total : 102 	Nb Ok total : 34.0 %
Nb OO : 0 	Nb OO total : 0 	Nb OO total : 0.0 %
Nb Er : 98 	Nb Er total : 198 	Nb Er total : 66.0 %
-------------
Programme :  4469
[['ADD', 'R3', 'R1'], ['ADD', 'R2', 'R2'], ['ADD', 'R0', 'R2'], 0, 0, 0, 0]
Nb Ok : 100 	Nb Ok total : 202 	Nb Ok total : 50.5 %
Nb OO : 0 	Nb OO total : 0 	Nb OO tota

### Analyses plus poussée

Sur la base de ce qui a été fait précédement, on peut pousser un peu plus une analyse particulière.

Exemple : donner le nombre maximum d'instructions exécutées (basée sur une analyse aléatoire).


In [ ]:
def ExecuteProcessusEtCompteInstruction(P):
    Tortue = P.copy()
    Lievre = P.copy()
    continuation = True
    NbInstruction = 0
    msgLievre = None
    msgTortue = None
    while continuation :
        msgTortue = ExecuteUneInstruction(Tortue)
        NbInstruction = NbInstruction+1
        if msgLievre is None :
            msgLievre = ExecuteUneInstruction(Lievre)
        if msgLievre is None :
            msgLievre = ExecuteUneInstruction(Lievre)
        if msgTortue == "Fin de programme" :
            return NbInstruction
        elif MemeEtat(Tortue,Lievre) :
            return -1
        elif msgTortue :
            return -1

In [80]:
def AnalyseNbMaxInstruction(NbMachines,NbDonnees) :
    print ("Analyse de ",NbMachines,"machines, chacune sur ",NbDonnees,"données, pour trouver le nombre maximum d'instructions exécutées")
    print("-------------")    
    NbMaxIns = 25 #pour limiter les affichages, 
    ProcessusAnalyse = {}
    for m in range(NbMachines) :
        NMachineAleatoire = randint(0,NbProgrammesPossibles)
        for i in range(NbDonnees):
            NDonneesAleatoires = randint(0,NbDonneesPossibles)
            MemoireAnalyse=GenereMachine(NMachineAleatoire,NDonneesAleatoires)
            initProcessus(ProcessusAnalyse,MemoireAnalyse)
            res=ExecuteProcessusEtCompteInstruction(ProcessusAnalyse)
            if res>NbMaxIns :
                NbMaxIns=res
                print ("Programme : ",NMachineAleatoire,"Donnee :",NDonneesAleatoires)
                print (GenereMachine(NMachineAleatoire,NDonneesAleatoires))
                print (NbMaxIns,"instructions exécutées.")
                print("-------------")

In [81]:
AnalyseNbMaxInstruction(100,10)

Analyse de  100 machines, chacune sur  10 données, pour trouver le nombre maximum d'instructions exécutées
-------------
Programme :  9257 Donnee : 14542
[['ADD', 'R3', 'R2'], ['ADD', 'R3', 'R2'], ['BNE', 1], 14, 6, 14, 0]
31 instructions exécutées.
-------------
Programme :  9257 Donnee : 169047
[['ADD', 'R3', 'R2'], ['ADD', 'R3', 'R2'], ['BNE', 1], 23, 2, 5, 5]
61 instructions exécutées.
-------------
Programme :  8525 Donnee : 590007
[['ADD', 'R3', 'R3'], ['ADD', 'R2', 'R0'], ['BNE', 0], 23, 5, 0, 18]
96 instructions exécutées.
-------------


ou le nombre moyen d'instructions exécutées (pour les programmes ok) :

In [82]:
def AnalyseNbMoyInstruction(NbMachines,NbDonnees) :
    print ("Analyse de ",NbMachines,"machines, chacune sur ",NbDonnees,"données, pour trouver le nombre moyen d'instructions exécutées")
    print("-------------")    
    NbTotIns = 0
    NbProgOk = 0
    ProcessusAnalyse = {}
    for m in range(NbMachines) :
        NMachineAleatoire = randint(0,NbProgrammesPossibles)
        for i in range(NbDonnees):
            NDonneesAleatoires = randint(0,NbDonneesPossibles)
            MemoireAnalyse=GenereMachine(NMachineAleatoire,NDonneesAleatoires)
            initProcessus(ProcessusAnalyse,MemoireAnalyse)
            res=ExecuteProcessusEtCompteInstruction(ProcessusAnalyse)
            if res>0 :
                NbTotIns=res+NbTotIns
                NbProgOk=NbProgOk+1
    print ("Nombre moyen d'instructions exécutées : ",NbTotIns/NbProgOk," pour ",NbProgOk,"programmes ok")


In [83]:
AnalyseNbMoyInstruction(100,10)

Analyse de  100 machines, chacune sur  10 données, pour trouver le nombre moyen d'instructions exécutées
-------------
Nombre moyen d'instructions exécutées :  4.818725099601593  pour  502 programmes ok


## Conclusion

### Récapitulatif

Le problème de l'arrêt peut donc être analysé de manière exhaustive pour une petite machine (2 instructions Add/BNE), ayant une petite mémoire (3 instructions, 4 données ; mots de 5 bits). L'analyse s'effectue en catégorisant les programmes selon qu'ils terminent normalement/correctement, qu'ils exécutent une instruction erronnée ou qu'ils bouclent à l'infini, resp. : qu'ils   exécutent la dernière instruction de la zone de programme sans que celle-ci n'engendre de saut (ni en avant, ni en arrière) ; qu'ils exécutent une instruction engendrant un saut hors de la zone de programme ; que l'exécution du programme passe plus d'une fois par le même état, i.e. que l'exécution du programme à deux vitesses différentes (une instruction à la fois, ou deux instructions à la suite à chaque fois) mène à un instant donnée au même état (cas de boucle infinie). 

Pour ces machines minimalistes et cette catégorisation des arrêts, l'analyse présente montre que :
* en moyenne, les machines correctes exécutent un peu plus de 5 instructions
* au maximum, les machines correctes exécutent 189 instructions
* parmi les 12167 programmes et 12758024192 machines différentes possibles, environ 50% des machines et programmes sont correctes (un peu moins), 40% terminent en erreur (un peu moins) et 10% bouclent à l'infini (un peu plus).

### Variantes

Pour la variante **8 bits** \[taille du mot mémoire] (à la place des 5 bits de la version de base) de ces machines :
* en moyenne, les machines correctes exécutent environ 25 instructions
* au maximum, les machines correctes peuvent exécuter plus de 1500 instructions
* parmi les programmes et machines différentes possibles, environ 50% des machines et programmes sont correctes (un peu moins), 40% terminent en erreur (un peu plus) et 10% bouclent à l'infini (un peu plus) 

Pour la variante **8 mots mémoire** (à la place des 7 mots mémoire de la version de base) de ces machines :
* en moyenne, les machines correctes exécutent environ 10 instructions
* au maximum, les machines correctes peuvent exécuter plus de 3000 instructions
* parmi les programmes et  machines différentes possibles, environ 40% des machines et programmes sont correctes, 40% terminent en erreur et 10% bouclent à l'infini (un peu plus).

Pour la variante **BEQ** (à la place de BNE de la version de base) de ces machines :
* en moyenne, les machines correctes exécutent 3 instructions
* au maximum, les machines correctes exécutent 9 instructions
* parmi les programmes et machines différentes possibles, sur des tests aléatoires de machine environ 99% des machines et programmes sont correctes (un peu moins), 1% terminent en erreur et moins de 1% bouclent à l'infini (mais un peu plus que 0%).

Pour la variante **sans erreur** (à la place de la version de base) de ces machines :
* en moyenne, les machines correctes exécutent un peu plus de 5 instructions
* au maximum, les machines correctes exécutent 189 instructions
* parmi les programmes et machines différentes possibles, environ 80% des machines et programmes sont correctes et terminent, 20% bouclent à l'infini.

Pour la variante **sûre**  (à la place de la version de base) de ces machines :
* en moyenne, les machines correctes exécutent 3 instructions
* au maximum, les machines correctes exécutent 48 instructions
* parmi les programmes et machines différentes possibles, environ 20% des machines et programmes sont correctes (un peu moins), 40% terminent avec une erreur de calcul (un peu plus), 30% terminent avec un saut invalide (un peu plus) et 10% bouclent à l'infini (un peu moins).

Pour la variante **normalisée** (14 ADD pour 7 BNE, c'est à dire 2/3 ADD pour 1/3 BNE, au lieu de 16 ADD pour 7 BNE) de ces machines :
* en moyenne, les machines correctes exécutent un peu moins de 6 instructions
* au maximum, les machines correctes peuvent exécuter 189 instructions
* parmi les programmes et machines différentes possibles, environ 40% des machines et programmes sont correctes (un peu plus), 40% terminent en erreur (un peu plus) et 10% bouclent à l'infini (un peu plus) 

### Rappels 

Pour effectuer une analyse, ou plusieurs analyses :

In [84]:
AnalyseMachine([["ADD","R3","R1"],["ADD","R0","R2"],["BNE",0],2,3,-1,0])

'Fin de programme'

In [85]:
AnalyseAlea(2,2)

Analyse de  2 machines, chacune sur  2 données
-------------
Programme :  440
[['ADD', 'R3', 'R0'], ['BNE', 3], ['ADD', 'R0', 'R0'], 0, 0, 0, 0]
Nb Ok : 0 	Nb Ok total : 0 	Nb Ok total : 0.0 %
Nb OO : 0 	Nb OO total : 0 	Nb OO total : 0.0 %
Nb Er : 2 	Nb Er total : 2 	Nb Er total : 100.0 %
-------------
Programme :  5809
[['ADD', 'R1', 'R3'], ['BNE', 6], ['ADD', 'R2', 'R2'], 0, 0, 0, 0]
Nb Ok : 0 	Nb Ok total : 0 	Nb Ok total : 0.0 %
Nb OO : 0 	Nb OO total : 0 	Nb OO total : 0.0 %
Nb Er : 2 	Nb Er total : 4 	Nb Er total : 100.0 %
-------------


In [86]:
AnalyseNbMaxInstruction(100,10)

Analyse de  100 machines, chacune sur  10 données, pour trouver le nombre maximum d'instructions exécutées
-------------
Programme :  8570 Donnee : 295904
[['ADD', 'R2', 'R3'], ['ADD', 'R0', 'R1'], ['BNE', 0], 0, 31, 0, 9]
96 instructions exécutées.
-------------


In [87]:
AnalyseNbMoyInstruction(100,10)

Analyse de  100 machines, chacune sur  10 données, pour trouver le nombre moyen d'instructions exécutées
-------------
Nombre moyen d'instructions exécutées :  5.859154929577465  pour  426 programmes ok


In [88]:
AnalyseDeA(1234,1236)

Analyse des machines
De : 1234
Jusqu'à : 1236
-------------
1234
Nb Ok : 1048576 	Nb Ok total : 1048576
Nb OO : 0 	Nb OO total : 0
Nb Er : 0 	Nb Er total : 0
-------------
1235
Nb Ok : 0 	Nb Ok total : 1048576
Nb OO : 1048576 	Nb OO total : 1048576
Nb Er : 0 	Nb Er total : 0
-------------
